examples/GAN/GAN.py

```diff
@@ -72,40 +73,89 @@ def get_optimizer(self):
+
+    If we don't care about inference during training, using tower_func is
+    not needed. Just calling model.build_graph directly is OK.
+    """
+    # Build the graph
+    cls.tower_func = TowerFuncWrapper(model.build_graph, model.get_inputs_desc())
+    with TowerContext('', is_training=True):
+        cls.tower_func(*input.get_input_tensors())
+    opt = model.get_optimizer()
+
+    # Define the training iteration
+    # by default, run one d_min after one g_min
+    with tf.name_scope('optimize'):
+        g_min = opt.minimize(model.g_loss, var_list=model.g_vars, name='g_op')
+        with tf.control_dependencies([g_min]):
+            d_min = opt.minimize(model.d_loss, var_list=model.d_vars, name='d_op')
+    cls.train_op = d_min
+
+
+def _build_multi_gan_trainer(cls, input, model, num_gpu=1):
+    assert num_gpu > 1
+    raw_devices = ['/gpu:{}'.format(k) for k in range(num_gpu)]
+
+    # Build the graph with multi-gpu replication
+    def get_cost(*inputs):
+        model.build_graph(*inputs)
+        return [model.d_loss, model.g_loss]
+
+    cls.tower_func = TowerFuncWrapper(get_cost, model.get_inputs_desc())
+    devices = [LeastLoadedDeviceSetter(d, raw_devices) for d in raw_devices]
+    cost_list = DataParallelBuilder.build_on_towers(
+        list(range(num_gpu)),
+        lambda: cls.tower_func(*input.get_input_tensors()),
+        devices)
+    # For simplicity, average the cost here. It might be faster to average the gradients
+    with tf.name_scope('optimize'):
+        d_loss = tf.add_n([x[0] for x in cost_list]) * (1.0 / num_gpu)
+        g_loss = tf.add_n([x[1] for x in cost_list]) * (1.0 / num_gpu)
+
+        opt = model.get_optimizer()
+        # run one d_min after one g_min
+        g_min = opt.minimize(g_loss, var_list=model.g_vars,
+                             colocate_gradients_with_ops=True, name='g_op')
+        with tf.control_dependencies([g_min]):
+            d_min = opt.minimize(d_loss, var_list=model.d_vars,
+                                 colocate_gradients_with_ops=True, name='d_op')
+    # Define the training iteration
+    cls.train_op = d_min
+
+
 class GANTrainer(TowerTrainer):
-    def __init__(self, input, model):
+
+    def __init__(self, input, model, num_gpu=1):
         """
         Args:
             input (InputSource):
             model (GANModelDesc):
         """
         super(GANTrainer, self).__init__()
         assert isinstance(model, GANModelDesc), model
-        inputs_desc = model.get_inputs_desc()
+
         # Setup input
-        cbs = input.setup(inputs_desc)
+        cbs = input.setup(model.get_inputs_desc())
         self.register_callback(cbs)
 
-        """
-        We need to set tower_func because it's a TowerTrainer,
-        and only TowerTrainer supports automatic graph creation for inference during training.
+        if num_gpu < 2:
+            _build_single_gan_trainer(self, input, model)
```

Comment:
Passing `self` around is an unusual design. Here we can just write them as methods of `GANTrainer`.

Reply:
You are correct. I also made a mistake with the StagingInput, which is fixed now.